### PR TITLE
Feature/collect all raw facts for network scan

### PR DESCRIPTION
**Pull request by bruno-fs**
Always collect raw fact names for network scan - even if they have a null value.

### CODE REVIEW

**Comment by bruno-fs** [comment]
@nicolearagao @gtanzillo please ignore https://github.com/quipucords/quipucords/pull/2150/commits/f1fc2b45ee63a701374de089b20ca175014eae1e for review purposes. I'll open a separate PR with just it

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml [hunk 1]
```diff
@@ -4,25 +4,17 @@
   set_fact:
     internal_host_started_processing_role: "redhat_release"
 
-- name: initialize redhat_release facts
```
Comment (nicolearagao) [inline quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:7]: Won't that cause any impacts on the way the report is printed in case there's no - "'stdout_lines' in internal_redhat_release"?

File: quipucords/scanner/network/utils.py [hunk 2]
```diff
@@ -102,3 +103,22 @@ def expand_hostpattern(hostpattern):
...
+def _yaml_load(path_obj):
+    return yaml.load(path_obj.open(), Loader=yaml.SafeLoader)
+
+
+def collect_all_fact_names():
+    """Collect all fact names set on ansible playbooks."""
+    network_runner_path = settings.BASE_DIR / "scanner/network/runner"
+    for playbook in network_runner_path.rglob("roles/*/tasks/main.yml"):
+        commands = _yaml_load(playbook)
+        for command in commands:
```
Comment (nicolearagao) [inline quipucords/scanner/network/utils.py:117]: Maybe replace commands with tasks?

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml [hunk 3]
```diff
@@ -4,25 +4,17 @@
...
-  set_fact:
-    redhat_release_name: ''
-    redhat_release_version: ''
-    redhat_release_release: ''
-  ignore_errors: yes
-
 - name: gather internal_redhat_release fact
   raw: rpm -q --queryformat "%{NAME}\n%{VERSION}\n%{RELEASE}\n" --whatprovides redhat-release
   register: internal_redhat_release
```
Comment (nicolearagao) [inline quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:9]: why did you change it to true instead of yes ?
It's a bit confussing but I guess Ansible standard is yes instead of true.
https://docs.ansible.com/ansible/latest/user_guide/playbooks_error_handling.html#

File: quipucords/fingerprinter/jboss_brms.py [hunk 4]
```diff
@@ -76,6 +76,14 @@ def classify_version_string(version_string):
     return None
 
 
+def _get_val_or_empty_list(data_dict, key):
```
Comment (nicolearagao) [inline quipucords/fingerprinter/jboss_brms.py:79]: Is this going to be called whenever we don't scan middleware ?

File: quipucords/scanner/network/test_sanity_check.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    inspect_data = _yaml_load(inspect_yml_path)
+    assert isinstance(inspect_data, list)
+    assert len(inspect_data) == 1
+    assert isinstance(inspect_data[0], dict)
+    assert (
+        inspect_data[0]["roles"][-1] == "host_done"
+    ), "host_done SHOULD be the last role"
+
+
+def test_check_all_roles_are_included(inspect_yml_path, path_to_playbooks):
```
Comment (nicolearagao) [inline quipucords/scanner/network/test_sanity_check.py:47]: This was pretty smart.

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml (hunk 1 above)
Comment (bruno-fs) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:7]: Mmm I think it will. Nice catch!

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml [hunk 6]
```diff
@@ -4,25 +4,17 @@
...
 
 - name: set redhat_release facts based on internal_redhat_release
   set_fact:
     redhat_release_name: "{{ internal_redhat_release['stdout_lines'][0] }}"
     redhat_release_version: "{{ internal_redhat_release['stdout_lines'][1] }}"
     redhat_release_release: "{{ internal_redhat_release['stdout_lines'][2] }}"
-  ignore_errors: yes
+  ignore_errors: true
   when:
     - "'stdout_lines' in internal_redhat_release"
```
Comment (bruno-fs) [inline quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:20]: @nicolearagao do you know if ansible `when` short circuits? If so I think we can still remove the initial setup and add this

```suggestion
    - internal_redhat_release is defined
    - "'stdout_lines' in internal_redhat_release"
    - "internal_redhat_release['stdout_lines']|length == 3"
```

PS: I shouldn't have removed the length check. probably did this by mistake

Comment (nicolearagao) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:20]: It's not short circuit evaluation. 
Why don't you do something like this? 
when: result.stdout | default('') == 'foo'

File: quipucords/fingerprinter/jboss_brms.py (hunk 4 above)
Comment (bruno-fs) [reply quipucords/fingerprinter/jboss_brms.py:79]: `detect_jboss_brms` will be called whether we scan middleware or not.
then it would use default dict getter for many keys with default set to `list()` or `set()`

This will be used later based on the premise that even if nothing is found an iterable will be there.
Changing the facts to always be filled with `None` breaks this behavior because the value `None` will take precedence over the default provided to the getter.

In other words...
```python
data_dict1 = {}
manifest_mfs1 = data_dict1.get(JBOSS_BRMS_MANIFEST_MF, set())
assert type(manifest_mfs1) == type(set())

data_dict2 = {JBOSS_BRMS_MANIFEST_MF: None}
manifest_mfs2 = data_dict2.get(JBOSS_BRMS_MANIFEST_MF, set())
assert type(manifest_mfs2) == type(None)
```

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml (hunk 6 above)
Comment (bruno-fs) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:20]: Not sure if I got it. Would this work if `internal_redhat_release` is `undefined`?

```suggestion
    - internal_redhat_release.stdout | default("\n\n\n")
```

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml (hunk 3 above)
Comment (nicolearagao) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:9]: awesome!!

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml (hunk 6 above)
Comment (nicolearagao) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:20]: I apologize, my suggestion wasn't clear.  if internal_redhat_release is undefined, it means that either this is not a red hat machine or that there was a problem while collecting this information. 
For the further I think it's still worth it to have a default value if anything goes wrong. I don't think calling set_fact twice to do it is the most efficient way. 
You could do:
set_fact:
    redhat_release_name: "{{ internal_redhat_release['stdout_lines'][0] | default(" ") }}"
    redhat_release_version: "{{ internal_redhat_release['stdout_lines'][1] | default(" ")}}"
    redhat_release_release: "{{ internal_redhat_release['stdout_lines'][2] | default(" ")}}"
The reason why this won't blow up if internal_redhat_release is undefined is due to the flag ignore_errors. In any other scenario it would.

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml [hunk 7]
```diff
@@ -4,25 +4,17 @@
...
-    - "internal_redhat_release['stdout_lines']|length == 3"
```
Comment (nicolearagao) [inline quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:28]: Why the length would be longer than 3? Probably if it raised an error, therefore o failed_when would be much more efficient than ignore_errors. 
I guess this calls for a refactor. hehe =P

Comment (bruno-fs) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml:28]: It's called "defensive programming". If the output don't have 3 lines, something is wrong.

File: quipucords/fingerprinter/jboss_fuse.py [hunk 8]
```diff
@@ -71,6 +68,10 @@ def get_version(eap_version):
     return new_versions
 
 
+def _get_val_or_empty_list(data_dict, key):
```
Comment (gtanzillo) [inline quipucords/fingerprinter/jboss_fuse.py:71]: Maybe put this in a common file with the other helper methods to DRY it up

**Review by gtanzillo — COMMENTED** [review]
@bruno-fs I did a first pass on this and it looks good. Since there are a lot of changes here, perhaps it would be good for you to do a walkthrough of the change.

File: quipucords/scanner/network/utils.py (hunk 2 above)
Comment (bruno-fs) [reply quipucords/scanner/network/utils.py:117]: change to glob

**Comment by bruno-fs** [comment]
we need a camayoc run before merging

**Comment by bruno-fs** [comment]
we need a camayoc run before merging

File: quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml (hunk 6 above)
Comment (bruno-fs) [reply quipucords/scanner/network/runner/roles/redhat_release/tasks/main.yml]: Setting a fact to " " can be problematic because the expected value for "empty" facts is "".
I could apply this suggestion with default to "" but I'm not confident this won't cause errors - it depends on 2 premises that might not be fulfilled (stdout_lines exists and has at least 3 lines).

Now I understood why this was implemented the way it was and I think I prefer this approach now. What do you think?

**Review by nicolearagao — APPROVED** [review]
LGTM. Great job!

File: quipucords/fingerprinter/jboss_fuse.py (hunk 8 above)
Comment (bruno-fs) [reply quipucords/fingerprinter/jboss_fuse.py]: Oh I didn't comment on this but I added a DRY solution in a project level `utils` module.

**Review by gtanzillo — APPROVED** [review]
Looks great! Thanks @bruno-fs

File: quipucords/fingerprinter/jboss_fuse.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bruno-fs) [reply quipucords/fingerprinter/jboss_fuse.py]: I replaced this one with the [default_getter](https://github.com/quipucords/quipucords/blob/96b821a58a7291c883d007c49e86f49418622bfb/quipucords/utils/default_getter.py)

Comment (bruno-fs) [reply quipucords/fingerprinter/jboss_fuse.py]: Replaced this function with the helper "default_getter" https://github.com/quipucords/quipucords/blob/96b821a58a7291c883d007c49e86f49418622bfb/quipucords/fingerprinter/jboss_fuse.py#L84-L101

**Comment by bruno-fs** [comment]
# CAMAYOC REPORT 

## API tests
```
(venv) [root@discovery-qpc-new camayoc]# make test-qpc-api -e PYTEST_OPTIONS='-vvv -ra --disable-warnings'
========================================================================================================= short test summary info ==========================================================================================================
SKIPPED [1] camayoc/tests/qpc/api/v1/reports/test_reports.py:319: Skipping sat6 run until Quipucords Issue #2039 is resolved                                                                                                               
SKIPPED [3] camayoc/tests/qpc/api/v1/scanjobs/test_pause_cancel_restart.py:23: Test is flaky. Skipping until Quipucords Issue #2040 resoloved                                                                                              
SKIPPED [3] camayoc/tests/qpc/api/v1/scanjobs/test_pause_cancel_restart.py:53: Test is flaky. Skipping until Quipucords Issue #2040 resoloved                                                                                              
SKIPPED [2] camayoc/tests/qpc/api/v1/scanjobs/test_run_scanjobs.py:31: Test is flaky. Skipping until Quipucords Issue #2040 resoloved                                                                                                      
XFAIL camayoc/tests/qpc/api/v1/reports/test_reports.py::test_report_content_consistency
  reason: Config file does not have dependent scan "rhel-7".
XFAIL camayoc/tests/qpc/api/v1/reports/test_reports.py::test_merge_reports_from_scanjob
  reason: Config file does not have dependent scans "non-rhel" or "rhel-7".
==================================================================================== 238 passed, 9 skipped, 2 xfailed, 7 warnings in 364.07s (0:06:04) =====================================================================================

```
